src/rgw/driver/sfs/types.cc

```diff
@@ -258,52 +323,75 @@ void MultipartUpload::abort(const DoutPrefixProvider *dpp) {
   objref.reset();
 }
 
-bool Bucket::want_new_version(const rgw_obj_key& key, ObjectRef obj) {
-  return !(key.instance.empty() || key.instance == obj->instance);
+bool Bucket::want_specific_version(const rgw_obj_key& key) {
+  return !key.instance.empty();
 }
 
 ObjectRef Bucket::get_or_create(const rgw_obj_key& key) {
-  std::lock_guard l(obj_map_lock);
-
   ObjectRef result;
-  try {
-    result = get_unmutexed(key.name);
-    if (want_new_version(key, result)) {
-      result->add_new_version(store);
+
+  auto maybe_result = Object::try_create_with_last_version_from_database_fetch(
+      store, key.name, info.bucket.bucket_id);
+
+  if (maybe_result == nullptr) {  // new object
+    result.reset(Object::create_commit_new_object(
+        key, store, info.bucket.bucket_id, &key.instance));
+    return result;
+  }
+
+  // an object exists with at least 1 version
+  if (want_specific_version(key) && maybe_result->instance == key.instance) {
+    // requested version happens to be the last version
+    result.reset(maybe_result);
+  } else if (want_specific_version(key) &&
+             maybe_result->instance != key.instance) {
+    // requested version is not last
+
+    auto specific_version_object = Object::try_create_from_database_fetch(
+        store, key.name, info.bucket.bucket_id, key.instance);
+
+    if (specific_version_object == nullptr) {
+      // requested version does not exist -> create it from last
+      // version object
```

Comment:
is this semantically correct? Should we be creating an object if the version does not exist at all?

Reply:
Not sure I understand that fully.

Before try_create_form_database_fetch we know:
- object exists
-  `>= 1` versions exists
- the requested version is not the last version

After we know in addition:
- the requested version exists or not

We are not creating an object here. That condition is at the beginning.


pre refactoring code for reference:
```
ObjectRef Bucket::get_or_create(const rgw_obj_key &key) {
  std::lock_guard l(obj_map_lock);
  ObjectRef obj = nullptr;
  auto new_object = true;
  auto create_new_version = true;
  auto it = objects.find(key.name);
  if (it != objects.end()) {
    obj = it->second;
    new_object = false;
    if (key.instance.empty() || key.instance == obj->instance) {
      create_new_version = false;
    } else {
      obj->instance = key.instance;
    }
  } else {
    obj = std::make_shared<Object>(key);
    objects[key.name] = obj;
  }
  obj->metadata_init(store, info.bucket.bucket_id, new_object, create_new_version);

  return obj;
}
```

Reply:
Regardless, the method screems for another round of refactoring, making every condition a db transaction/single sql statement